From: Christoph Hellwig <hch@lst.de>
Date: Thu, 21 Aug 2014 11:09:17 -0500
Subject: [PATCH] nfs: cap request size to fit a kmalloced page array
Git-commit: 2e11f8296d22134c4fca7eb022eea2b09facd307
Patch-mainline: v3.18
References: bnc#898675

pNFS servers may return arbitrarily large layouts.  Trim back the I/O size
to one that we can at least allocate the page array for.

Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Acked-by: NeilBrown <neilb@suse.de>

---
 fs/nfs/pagelist.c |    8 ++++++++
 1 file changed, 8 insertions(+)

--- linux-3.12-SLE12.orig/fs/nfs/pagelist.c
+++ linux-3.12-SLE12/fs/nfs/pagelist.c
@@ -291,6 +291,14 @@ bool nfs_generic_pg_test(struct nfs_page
 	if (desc->pg_bsize < PAGE_SIZE)
 		return 0;
 
+	/*
+	 * Limit the request size so that we can still allocate a page array
+	 * for it without upsetting the slab allocator.
+	 */
+	if (((desc->pg_count + req->wb_bytes) >> PAGE_SHIFT) *
+			sizeof(struct page) > PAGE_SIZE)
+		return false;
+
 	return desc->pg_count + req->wb_bytes <= desc->pg_bsize;
 }
 EXPORT_SYMBOL_GPL(nfs_generic_pg_test);
